File: libs/api/src/main/java/com/example/api/ApiResponse.java
package com.example.api;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.IOException;

import retrofit2.Response;

public class ApiResponse<T> {
	private final int statusCode;

	@Nullable
	private final T body;

	@Nullable
	private final String errorMessage;

	public ApiResponse(@NonNull Throwable error) {
		statusCode = 500;
		body = null;
		errorMessage = error.getMessage();
	}

	ApiResponse(@NonNull Response<T> response) {
		statusCode = response.code();

		if (response.isSuccessful()) {
			body = response.body();
			errorMessage = null;
		} else {
			String message = null;

			if (response.errorBody() != null) {
				try {
					message = response.errorBody().string();
				} catch (IOException ignored) {

				}
			}

			if (message == null || message.isEmpty()) {
				message = response.message();
			}

			errorMessage = message;
			body = null;
		}
	}

	public boolean isSuccessful() {
		return statusCode >= 200 && statusCode < 300;
	}

	@Nullable
	public T getBody() {
		return body;
	}
}

File: libs/api/src/main/kotlin/com/example/api/ApiResponse.kt
package com.example.api

import retrofit2.Response
import java.io.IOException

class ApiResponse<T> {
  private val statusCode: Int

  val body: T?

  private val errorMessage: String?

  val isSuccessful: Boolean
    get() = statusCode in 200..399

  constructor(error: Throwable) {
    statusCode = 500
    body = null
    errorMessage = error.message
  }

  internal constructor(response: Response<T>) {
    statusCode = response.code()

    if (response.isSuccessful) {
      body = response.body()
      errorMessage = null
    } else {
      var message: String? = null

      if (response.errorBody() != null) {
        try {
          message = response.errorBody()!!.string()
        } catch (ignored: IOException) {
        }
      }

      if (message == null || message.isEmpty()) {
        message = response.message()
      }

      errorMessage = message
      body = null
    }
  }
}
